src/robot/libdocpkg/specbuilder.py

```diff
@@ -45,39 +46,60 @@ def _parse_spec(self, path):
             root = ET.parse(source).getroot()
         if root.tag != 'keywordspec':
             raise DataError("Invalid spec file '%s'." % path)
+        version = root.get('specversion')
+        if version != '3':
+            raise DataError("Invalid spec file version '%s'. "
+                            "Robot Framework >= 4.0 requires specversion 3." % version)
         return root
 
-    def _get_scope(self, spec):
-        # RF >= 3.2 has "scope" attribute w/ value 'GLOBAL', 'SUITE, or 'TEST'.
-        if 'scope' in spec.attrib:
-            return spec.get('scope')
-        # RF < 3.2 has "scope" element. Need to map old values to new.
-        scope = spec.find('scope').text
-        return {'': 'GLOBAL',          # Was used with resource files.
-                'global': 'GLOBAL',
-                'test suite': 'SUITE',
-                'test case': 'TEST'}[scope]
-
-    def _get_named_args(self, spec):
-        # RF >= 3.2 has "namedargs" attribute w/ value 'true' or 'false'.
-        namedargs = spec.get('namedargs')
-        if namedargs == 'true':
-            return True
-        if namedargs == 'false':
-            return False
-        # RF < 3.2 has "namedargs" element with text 'yes' or 'no'.
-        namedargs = spec.find('namedargs').text
-        return namedargs == 'yes'
-
     def _create_keywords(self, spec, path):
         return [self._create_keyword(elem) for elem in spec.findall(path)]
 
     def _create_keyword(self, elem):
         # "deprecated" attribute isn't read because it is read from the doc
         # automatically. That should probably be changed at some point.
         return KeywordDoc(name=elem.get('name', ''),
-                          args=[a.text for a in elem.findall('arguments/arg')],
+                          args=self._create_arguments(elem),
                           doc=elem.find('doc').text or '',
                           tags=[t.text for t in elem.findall('tags/tag')],
                           source=elem.get('source'),
                           lineno=int(elem.get('lineno', -1)))
+
+    def _create_arguments(self, elem):
+        spec = ArgumentSpec()
+        setters = {
+            ArgInfo.POSITIONAL_ONLY: spec.positional_only.append,
+            ArgInfo.POSITIONAL_OR_NAMED: spec.positional_or_named.append,
+            ArgInfo.VAR_POSITIONAL: lambda value: setattr(spec, 'var_positional', value),
+            ArgInfo.NAMED_ONLY: spec.named_only.append,
+            ArgInfo.VAR_NAMED: lambda value: setattr(spec, 'var_named', value),
+        }
+        for arg in elem.findall('arguments/arg'):
+            name = self._get_name(arg)
+            if name:
+                setters[arg.get('kind')](name)
+                spec.defaults.update(self._get_default(arg, name))
+                spec.type = self._get_updated_types(arg, name, spec)
```

Comment:
The above with the needed helpers is pretty complicated. This would in my opinion be a lot simpler considering it needs no helpers:

```python
name = arg.find('name')
if not name:
    continue
name = name.text
default = arg.find('default')
if default:
    spec.defaults[name] = default.text
type = arg.find('type')
if type:
    if not spec.types:
        spec.types = {}
    spec.types[name] = type.text
```

Could obviously use `name_elem` instead of `name` to avoid using the same variable name.

Notice also that the original code has a bug: `spec.type` should be `spec.types`.

Reply:
i changed it.

But you forgot to write the kind with the setter and 
it is essential to check if xml_elem is `None` because also a found xml_element is falsly.
i don't know why, but that is it.